File: wordleCalculator.py
import time
import os
import pickle
from functools import wraps

def cache_storage(storage_path):
    def decorator(func):
        cache_file = os.path.join(storage_path, f"{func.__name__}_cache.pkl")

        @wraps(func)
        def wrapper(*args, **kwargs):
            # Use a tuple of arguments and keyword arguments as the cache key
            key = tuple(args[0]) if args else None 
            
            # Create the storage directory if it doesn't exist
            os.makedirs(storage_path, exist_ok=True)

            # Check if the cache file exists
            if os.path.exists(cache_file):
                with open(cache_file, 'rb') as file:
                    cache = pickle.load(file)
            else:
                cache = {}

            # Check if the result is already in the cache
            if key in cache:
                # print(f"Cache hit for {func.__name__}{key}")
                return cache[key]

            # If not in cache, compute the result and store it
            result = func(*args, **kwargs)
            cache[key] = result
            # print(f"Caching result for {func.__name__}{key}")

            # Save the updated cache to the file
            with open(cache_file, 'wb') as file:
                pickle.dump(cache, file)

            return result
    
        def remove_cache(*args, **kwargs):
            # Remove cache based on specified criteria
            key_to_remove = args[0] if args else None

            if key_to_remove:
                try:
                    with open(cache_file, 'rb') as file:
                        cache = pickle.load(file)
                        if key_to_remove in cache:
                            del cache[key_to_remove]
                            with open(cache_file, 'wb') as file:
                                pickle.dump(cache, file)
                            print(f"Cache for {func.__name__}{key_to_remove} removed.")
                        else:
                            print(f"No cache found for {func.__name__}{key_to_remove}.")
                except FileNotFoundError:
                    print("Cache file not found.")

        wrapper.remove_cache = remove_cache
        return wrapper

    return decorator

# returns the result of a guess answer pair
def wordleResult(guess, answer):
    result = [0,0,0,0,0]
    i = 0
    str = ""
    for letter in guess:
        # case for green letter
        if letter == answer[i]:
            result[i] = 2
        # case for yellow or grey letter
        else:
            str += answer[i]
        i += 1
    i = 0
    for letter in guess:
        if result[i] !=2:
            # loop through again and look at non greens
            j = 0
            for l in str:
                if letter == l:
                    result[i] = 1
                    str = str[:j] + str[j+1:]
                j+=1
        i += 1
    return (result[0], result[1], result[2], result[3], result[4])

# returns true if an answer is possible given a guess and result
def isPossibleAnswer(answer, guess, result):
    return wordleResult(guess, answer) == result

# returns all possible answers
def getPossibleAnswers(answerList, guess, result):
    possible = []
    for answer in answerList:
        if isPossibleAnswer(answer, guess, result):
            possible += [answer]
    return possible

# scores all answers
def scoreAnswers(validGuess, validAnswer):
    # print("Number of possible Answers:", len(validAnswer))
    guessNumbers = []
    i = 0
    ret = []
    maxScore = -1
    max = validGuess[0]
    # for every guess assume every answer and see how many answers it would eliminate
    # sum total eliminated answers for score
    for guess in validGuess:
        guessNumbers += [0]
        for answer in validAnswer:
            result = wordleResult(guess, answer)
            for ans in validAnswer:
                if not isPossibleAnswer(ans, guess, result):
                    guessNumbers[i] += 1
        ret += [(guess[0][0] + guess[1][0] + guess[2][0] + guess[3][0] + guess[4][0], guessNumbers[i])]
        if guessNumbers[i] > maxScore:
            maxScore = guessNumbers[i]
            max = guess
        i += 1
    # print(maxScore)
    return max

# eliminates guesses that are deemed bad due to letter frequency
def widdle(validAnswers, validGuesses, results):
    letterFreq = {'a': 0, 'b': 0, 'c': 0, 'd': 0, 'e': 0, 'f': 0, 'g': 0, 'h': 0, 'i': 0, 'j': 0, 'k': 0, 'l': 0,
               'm': 0, 'n': 0, 'o': 0, 'p': 0, 'q': 0, 'r': 0, 's': 0, 't': 0, 'u': 0, 'v': 0, 'w': 0,
               'x': 0, 'y': 0, 'z': 0}
    for answer in validAnswers:
        for letter in answer:
            letterFreq[letter] += 1
    grey = []
    for entry in results:
        word = entry[0]
        result = entry[1]
        i = 0
        str = ''
        for number in result:
            if number == 0 and word[i] not in str:
                grey += word[i]
            else:
                str = str + word[i]
            i+=1
    i=0
    for letter, val in letterFreq.items():
        if val == 0:
            grey += [letter]
        i+=1
    ret = []
    while(len(grey) > 18):
        grey.remove(grey[len(grey)-1])
    for word in validGuesses:
        bool = True
        for letter in grey:
            if letter in word:
                bool = False
                break
        if bool:
            ret += [word]
    # print("Widdled! Down to", len(ret), "from",len(validGuess))
    return ret

# eliminates more strictly based on letter frequency in remaining answers
# this function sometimes breaks im not sure why so im not using it anymore
def widdleMore(validAnswers, validGuesses):
    letterFreq = {'a': 0, 'b': 0, 'c': 0, 'd': 0, 'e': 0, 'f': 0, 'g': 0, 'h': 0, 'i': 0, 'j': 0, 'k': 0, 'l': 0,
                  'm': 0, 'n': 0, 'o': 0, 'p': 0, 'q': 0, 'r': 0, 's': 0, 't': 0, 'u': 0, 'v': 0, 'w': 0,
                  'x': 0, 'y': 0, 'z': 0}
    for answer in validAnswers:
        for letter in answer:
            letterFreq[letter] += 1
    least = 13000
    worst = ''
    for letter, val in letterFreq.items():
        if val < least and val != 0:
            least = val
            worst = letter
    ret = []
    for word in validGuesses:
        if worst not in word:
            ret += [word]
    used = [worst]
    while len(ret) > 700:
        least = 13000
        nworst = ''
        for letter, val in letterFreq.items():
            if val < least and val != 0 and letter not in used:
                least = val
                nworst = letter
        ret2 = []
        for word in ret:
            if nworst not in word:
                ret2 += [word]
        used += [nworst]
        if len(ret2) == 0:
            return ret
        ret = ret2
    # print("Widdled More! Down to", len(ret), "from",len(validGuess))
    return ret

def scoreAnswers2(validGuess, validAnswer):
    max_score = 0
    max_word = ""

    for word in validGuess:
        s = set()
        for answer in validAnswer:
            s.add(wordleResult(word, answer))
        if len(s) > max_score:
            max_score = len(s)
            max_word = word
        elif len(s) == max_score and word in validAnswer:
            max_score = len(s)
            max_word = word
    return max_word

def scoreAnswers3(validGuess, validAnswer):
    scores = []

    for word in validGuess:
        s = set()
        for answer in validAnswer:
            s.add(wordleResult(word, answer))
        scores.append((word, len(s)))

    # Sort the scores in descending order
    scores.sort(key=lambda x: (x[1], x[0] in validAnswer), reverse=True)

    # Return the top 3 words
    top_words = [word for word, _ in scores[:2]]
    return top_words

def runCalc(validAnswer, validGuesses, secretAnswer, startingWord):
    # print("Guessing:", startingWord)
    resultsArr = [(startingWord, wordleResult(startingWord, secretAnswer))]
    turns = 1

    while resultsArr[-1][1] != (2,2,2,2,2):
        currGuess = getNextWord2(tuple(resultsArr), validGuesses, validAnswer)
        # print("Guessing: " + currGuess)
        resultsArr += [(currGuess, wordleResult(currGuess, secretAnswer))]
        turns += 1
    return turns

def runCalcFromTree(validAnswer, validGuesses, secretAnswer):
    # print("Guessing:", startingWord)
    resultsArr = []
    turns = 0
    while len(resultsArr) == 0 or resultsArr[-1][1] != (2,2,2,2,2):
        info = treeTreversal(resultsArr, validGuesses, validAnswer)
        currGuess = info[0]
        print("Guessing: " + currGuess)
        resultsArr += [(currGuess, wordleResult(currGuess, secretAnswer))]
        turns += 1
    return turns

@cache_storage(storage_path="cache/old")
def getNextWord(resultsArr, validGuess, validAnswer):
    validAnswer = list(validAnswer)
    validGuess = list(validGuess)
    key = []
    for info in resultsArr:
        key += [info[1]]
        result = info[1]
        guess = str(info[0]).lower()
        validAnswer = getPossibleAnswers(validAnswer, guess, result)

    optimalGuesses = validGuess
    optimalGuesses = widdle(validAnswer, validGuess, resultsArr)
    if (len(validAnswer) > 3):
        bestScore = scoreAnswers(optimalGuesses, validAnswer)
    else:
        bestScore = scoreAnswers(validAnswer, validAnswer)
    guess = bestScore
    return guess

# @cache_storage(storage_path="cache/tree")
def getNextWord2(resultsArr, validGuess, validAnswer):
    validAnswer = list(validAnswer)
    validGuess = list(validGuess)
    key = []
    for info in resultsArr:
        key += [info[1]]
        result = info[1]
        guess = str(info[0]).lower()
        validAnswer = getPossibleAnswers(validAnswer, guess, result)

    optimalGuesses = validGuess
    optimalGuesses = widdle(validAnswer, validGuess, resultsArr)
    if (len(validAnswer) > 3):
        bestScore = scoreAnswers2(optimalGuesses, validAnswer)
    else:
        bestScore = scoreAnswers2(validAnswer, validAnswer)

    guess = bestScore
    return guess

def getNextWords(resultsArr, validGuess, validAnswer):
    validAnswer = list(validAnswer)
    validGuess = list(validGuess)
    key = []
    for info in resultsArr:
        key += [info[1]]
        result = info[1]
        guess = str(info[0]).lower()
        validAnswer = getPossibleAnswers(validAnswer, guess, result)

    optimalGuesses = validGuess
    # optimalGuesses = widdle(validAnswer, validGuess, resultsArr)
    if (len(validAnswer) > 3):
        bestScore = scoreAnswers3(optimalGuesses, validAnswer)
    else:
        bestScore = [scoreAnswers2(validAnswer, validAnswer)]

    guess = bestScore
    return guess

def startingWord():
    
    # get list of words and format them
    # get starting outcomes for roate
    validGuessFile = open('wordle/validGuess.txt', 'r')
    validAnswerFile = open('wordle/validAnswer.txt', 'r')
    validGuess = tuple(validGuessFile.read().split('\n'))
    validAnswer = validAnswerFile.read().split('\n')
    validAnswerFile.close()
    validGuessFile.close()

    
    sword = ""
    while sword not in validGuess:
        sword = input("Starting Word: ")
    print()

    average = 0
    i = 0
    start = time.time()
    for answer in validAnswer:
        i += 1
        print(f"Word: {answer}")
        last = runCalc(tuple(validAnswer), validGuess, answer, sword)
        average += last
        print(f"Guesses: {last}")
        print()

    end = time.time()

    sfile = open("starting_words.txt", "a+")
    sfile.write("\n")
    sfile.write(sword)
    sfile.write("\n\n")
    sfile.write(f"Total Time: {end - start}\n")
    average = average / (i * 1.0)
    sfile.write(f'Average Guesses: {average}\n')
    sfile.write(f'Average Time: {(end - start) / (i * 1.0)}\n')
    sfile.close()

def runAll():
    
    # get list of words and format them
    # get starting outcomes for roate
    validGuessFile = open('wordle/validGuess.txt', 'r')
    validAnswerFile = open('wordle/validAnswer.txt', 'r')
    validGuess = tuple(validGuessFile.read().split('\n'))
    validAnswer = validAnswerFile.read().split('\n')
    validAnswerFile.close()
    validGuessFile.close()

    average = 0
    i = 0
    start = time.time()
    sword = ""
    for answer in validAnswer:
        i += 1
        print(f"Word: {answer}")
        last = runCalcFromTree(validAnswer, validGuess, answer)
        average += last
        print(f"Guesses: {last}")
        print()

    end = time.time()

    sfile = open("starting_words.txt", "a+")
    sfile.write("\n")
    sfile.write(sword)
    sfile.write("\n\n")
    sfile.write(f"Total Time: {end - start}\n")
    average = average / (i * 1.0)
    sfile.write(f'Average Guesses: {average}\n')
    sfile.write(f'Average Time: {(end - start) / (i * 1.0)}\n')
    sfile.close()

@cache_storage("wordle/cache/tree")
def treeTreversal(resultsArr, validGuess, validAnswer):
    print("Running")
    # if len(resultsArr) == 0 and sword != "":
    #     currGuesses = [sword]
    # else:
    return (getNextWord2(resultsArr, validGuess, validAnswer), 0)
    currGuesses = getNextWords(tuple(resultsArr), validGuess, validAnswer)
    validAnswer2 = list(validAnswer)
    key = []
    for info in resultsArr:
        key += [info[1]]
        result = info[1]
        guess = str(info[0]).lower()
        validAnswer2 = getPossibleAnswers(validAnswer2, guess, result)
    min_total = -1
    min_guess = ""
    for currGuess in currGuesses:
        possibleResults = set()
        for answer in validAnswer2:
            possibleResults.add(wordleResult(currGuess, answer))
        total = 0
        for result in possibleResults:
            resultsArr += [(currGuess, result)]
            if result == (2,2,2,2,2):
                f = open("test.txt", "a")
                f.write(currGuess + "\n")
                f.close()
                total += len(resultsArr)
            else:
                total += treeTreversal(resultsArr, validGuess, validAnswer)[1]
            resultsArr.pop()
        if total < min_total or min_total == -1:
            min_total = total
            min_guess = currGuess
    return (min_guess, min_total)

def startTreeTreversal():
    # get list of words and format them
    # get starting outcomes for roate
    validGuessFile = open('wordle/validGuess.txt', 'r')
    validAnswerFile = open('wordle/validAnswer.txt', 'r')
    validGuess = tuple(validGuessFile.read().split('\n'))
    validAnswer = validAnswerFile.read().split('\n')
    validAnswerFile.close()
    validGuessFile.close()

    sword = ""
    # while sword not in validGuess:
    #     sword = input("Starting Word: ")
    # print()

    average = 0
    i = 0
    start = time.time()

    total = treeTreversal([], validGuess, validAnswer, sword)

    end = time.time()

    print("Time:", end - start)
    print(total[1] / len(validAnswer))

    # sfile = open("starting_words.txt", "a+")
    # sfile.write("\n")
    # sfile.write(sword)
    # sfile.write("\n\n")
    # sfile.write(f"Total Time: {end - start}\n")
    # average = average / (i * 1.0)
    # sfile.write(f'Average Guesses: {average}\n')
    # sfile.write(f'Average Time: {(end - start) / (i * 1.0)}\n')
    # sfile.close()



# cached_my_function = cache_function_output(storage_path="/test")(getNextWord)
# cached_my_function.remove_cache((('trace', (0, 0, 2, 1, 0)), ('cable', (2, 1, 0, 1, 0))))

if __name__ == "__main__":
    validGuessFile = open('wordle/validGuess.txt', 'r')
    validAnswerFile = open('wordle/validAnswer.txt', 'r')
    validGuess = tuple(validGuessFile.read().split('\n'))
    validAnswer = validAnswerFile.read().split('\n')
    validAnswerFile.close()
    validGuessFile.close()
    
    print(treeTreversal([('SALET', (0, 1, 1, 0, 0)), ('DRINK', (1, 0, 1, 0, 0))], validGuess, validAnswer))

File: static/wordleCalculator.js
// Implement a simple LRU (Least Recently Used) cache
class LRUCache {
    constructor(capacity) {
      this.capacity = capacity;
      this.cache = new Map();
    }
  
    get(key) {
      if (this.cache.has(key)) {
        const value = this.cache.get(key);
        // Move the accessed item to the end of the Map to indicate it's recently used
        this.cache.delete(key);
        this.cache.set(key, value);
        return value;
      }
      return null;
    }
  
    put(key, value) {
      if (this.cache.has(key)) {
        this.cache.delete(key);
      } else if (this.cache.size >= this.capacity) {
        // If the cache is full, remove the least recently used (first item)
        this.cache.delete(this.cache.keys().next().value);
      }
      this.cache.set(key, value);
    }
  }
  

const wordleResultCache = new LRUCache(1000);

function wordleResult(guess, answer) {
    const cacheKey = JSON.stringify({ guess, answer });
    if (wordleResultCache.has(cacheKey)) {
        return wordleResultCache.get(cacheKey);
    }
    const result = [0, 0, 0, 0, 0];
    const str = new Map();
    let i = 0;

    for (const letter of guess) {
        if (letter === answer[i]) {
        result[i] = 2; // case for green letter
        } else {
        str.set(answer[i], (str.get(answer[i]) || 0) + 1); // case for yellow or grey letter
        }
        i += 1;
    }

    i = 0;

    for (const letter of guess) {
        if (result[i] !== 2) {
        // loop through again and look at non-greens
        if (str.get(letter) > 0) {
            // if the letter is not green and is in the answer, subtract from our dictionary of the non-green letters
            result[i] = 1;
            str.set(letter, str.get(letter) - 1);
        }
        }
        i += 1;
    }

    return [result[0], result[1], result[2], result[3], result[4]];
}

console.log(wordleResult("APPLE", "ABCDE")); // Example usage

// Helper function to check if a given answer is a possible answer based on the guess and result
function isPossibleAnswer(answer, guess, result) {
    return JSON.stringify(wordleResult(guess, answer)) === JSON.stringify(result);
}
  
// Returns all possible answers
function getPossibleAnswers(answerList, guess, result) {
    const possible = [];
    for (const answer of answerList) {
        if (isPossibleAnswer(answer, guess, result)) {
        possible.push(answer);
        }
    }
    return possible;
}

// Scores all answers
function scoreAnswers(validGuess, validAnswer) {
    let maxScore = 0;
    let max = validGuess[0];

    // For every guess, assume every answer and see how many answers it would eliminate
    // Sum total eliminated answers for the score
    for (const guess of validGuess) {
        let currScore = 0;
        for (const answer of validAnswer) {
        const result = wordleResult(guess, answer);
        for (const ans of validAnswer) {
            if (!isPossibleAnswer(ans, guess, result)) {
            currScore += 1;
            }
        }
        }
        if (currScore > maxScore) {
        max = guess;
        maxScore = currScore;
        }
    }
    return max;
}

function widdle(validAnswers, validGuesses, results) {
    const letterFreq = new Map();
    const grey = new Set();
  
    // Calculate the letter frequency in validAnswers
    for (const answer of validAnswers) {
      for (const letter of answer) {
        letterFreq.set(letter, (letterFreq.get(letter) || 0) + 1);
      }
    }
  
    // Determine the grey letters based on results
    for (const entry of results) {
      const word = entry[0];
      const result = entry[1];
      let str = '';
      for (let i = 0; i < result.length; i++) {
        if (result[i] === 0 && !str.includes(word[i])) {
          grey.add(word[i]);
        } else {
          str += word[i];
        }
      }
    }
  
    // Add letters with zero frequency to grey
    for (const letter of letterFreq.keys()) {
      if (!letterFreq.get(letter)) {
        grey.add(letter);
      }
    }
  
    // Remove extra letters from grey if it has more than 18 elements
    while (grey.size > 18) {
      grey.delete(Array.from(grey)[grey.size - 1]);
    }
  
    // Filter validGuesses based on grey letters
    const ret = validGuesses.filter(word => {
      for (const letter of grey) {
        if (word.includes(letter)) {
          return false;
        }
      }
      return true;
    });
    return ret;
}
  
function widdleMore(validAnswers, validGuesses) {
    const letterFreq = new Map();
  
    // Calculate the letter frequency in validAnswers
    for (const answer of validAnswers) {
      for (const letter of answer) {
        letterFreq.set(letter, (letterFreq.get(letter) || 0) + 1);
      }
    }
  
    // Find the least frequent letter
    let least = 13000;
    let worst = '';
    for (const [letter, val] of letterFreq.entries()) {
      if (val < least && val !== 0) {
        least = val;
        worst = letter;
      }
    }
  
    // Filter validGuesses based on the least frequent letter
    const ret = validGuesses.filter(word => !word.includes(worst));
  
    // Remove the least frequent letter and filter validGuesses further until the desired length is reached
    const used = new Set([worst]);
    while (ret.length > 700) {
      least = 13000;
      let nworst = '';
      for (const [letter, val] of letterFreq.entries()) {
        if (val < least && val !== 0 && !used.has(letter)) {
          least = val;
          nworst = letter;
        }
      }
      const ret2 = ret.filter(word => !word.includes(nworst));
      used.add(nworst);
      if (ret2.length === 0) {
        return ret;
      }
      ret = ret2;
    }
  
    return ret;
}
function rankScores(scoredWords) {
    const scoresDict = {};
    const scoresList = [];
  
    for (const row of scoredWords) {
      let currScore = parseFloat(row[1]);
      while (scoresList.includes(currScore)) {
        currScore += 1 / scoresList.length;
      }
      scoresList.push(currScore);
      scoresDict[currScore] = row[0];
    }
  
    const score = Math.max(...scoresList);
    return scoresDict[score];
}

const getNextWordCache = new Map();

function getNextWord(results) {
    if (getNextWordCache.has(results)) {
        return getNextWordCache.get(results);
    }

    const validGuessFile = 'wordle/validGuess.txt';
    const validAnswerFile = 'wordle/validAnswer.txt';
    const firstResultFile = 'wordle/beginning_roate.txt';

    

    const fs = require('fs');

    let firstResult = fs.readFileSync(firstResultFile, 'utf8').split('\n');
    let validGuess = fs.readFileSync(validGuessFile, 'utf8').split('\n');
    let validAnswer = fs.readFileSync(validAnswerFile, 'utf8').split('\n');

    for (const info of results) {
        const result = info[1];
        const guess = info[0].toLowerCase();
        validAnswer = getPossibleAnswers(validAnswer, guess, result);
        if (results.length === 1 && validAnswer.length > 10) {
        const line = firstResult[result[4] + result[3] * 3 + result[2] * 9 + result[1] * 27 + result[0] * 81];
        const guess = line.slice(-5);
        cache.set(results, guess);
        return guess;
        }
    }

    let optimalGuesses = validGuess;
    let bestScore;
    if (validAnswer.length > 3) {
        bestScore = scoreAnswers(optimalGuesses, validAnswer);
    } else {
        bestScore = scoreAnswers(validAnswer, validAnswer);
    }

    const guess = bestScore;
    cache.set(results, guess);
    return guess;
}
